src/context/providers.tsx

```diff
@@ -1,139 +1,236 @@
+const ProvidersContext = createContext<Nullable<Providers>>(null);
 
-type ProviderProviderProps = {
+type ProvidersContextProps = {
   children: React.ReactNode;
 };
 
-/**
- * Returns two blockchain providers.
- *
- * The infura provider is always available, regardless of whether or not a
- * wallet is connected.
- *
- * The web3 provider, however, is based on the conencted and wallet and will
- * therefore be null if no wallet is connected.
- */
-export function ProvidersProvider({children}: ProviderProviderProps) {
-  const {provider} = useWallet();
+export function ProvidersContextProvider({children}: ProvidersContextProps) {
   const {network} = useNetwork();
+  const {provider} = useWallet();
+  const apiProvider = useSpecificProvider(network);
 
-  const [infuraProvider, setInfuraProvider] = useState<Providers['infura']>(
-    new InfuraProvider(NW_ARB, infuraApiKey)
+  // given that nothing should work on unsupported networks,
+  // asserting that apiProvider is always non null.
+  const contextValue = useMemo(
+    () => ({api: apiProvider!, web3: provider}),
+    [apiProvider, provider]
   );
 
-  useEffect(() => {
-    setInfuraProvider(getInfuraProvider(network));
-  }, [network]);
-
   return (
-    <ProviderContext.Provider
-      // TODO: remove casting once useSigner has updated its version of the ethers library
-      value={{infura: infuraProvider, web3: (provider as Web3Provider) || null}}
-    >
+    <ProvidersContext.Provider value={contextValue}>
       {children}
-    </ProviderContext.Provider>
+    </ProvidersContext.Provider>
   );
 }
 
-function getInfuraProvider(network: SupportedNetworks) {
-  // NOTE Passing the chainIds from useWallet doesn't work in the case of
-  // arbitrum and arbitrum-goerli. They need to be passed as objects.
-  // However, I have no idea why this is necessary. Looking at the ethers
-  // library, there's no reason why passing the chainId wouldn't work. Also,
-  // I've tried it on a fresh project and had no problems there...
-  // [VR 07-03-2022]
-  if (network === 'arbitrum') {
-    return new InfuraProvider(NW_ARB, infuraApiKey);
-  } else if (network === 'arbitrum-test') {
-    return new InfuraProvider(NW_ARB_GOERLI, infuraApiKey);
-  } else if (network === 'mumbai' || network === 'polygon') {
-    return new JsonRpcProvider(CHAIN_METADATA[network].rpc[0], {
-      chainId: CHAIN_METADATA[network].id,
-      name: translateToNetworkishName(network),
-      ensAddress:
-        LIVE_CONTRACTS[
-          translateToNetworkishName(network) as sdkSupportedNetworks
-        ].ensRegistry,
-    });
-  } else {
-    return new InfuraProvider(CHAIN_METADATA[network].id, infuraApiKey);
+/* CONTEXT CONSUMER ========================================================= */
+
+export function useProviders(): NonNullable<Providers> {
+  const context = useContext(ProvidersContext);
+
+  // Check if context is defined
+  if (context === undefined) {
```

Comment:
We should probably check for `null` values as well
```suggestion
  if (context == null) {
```